test/env/env.go

```diff
@@ -0,0 +1,142 @@
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	webhooktesting "github.com/jetstack/cert-manager/cmd/webhook/app/testing"
+	"github.com/onsi/ginkgo"
+	ginkgoconfig "github.com/onsi/ginkgo/config"
+	"github.com/onsi/ginkgo/reporters"
+	"github.com/onsi/gomega"
+	extapi "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+
+	cmpapi "github.com/cert-manager/policy-approver/apis/policy/v1alpha1"
+)
+
+func init() {
+	// Turn on verbose by default to get spec names
+	ginkgoconfig.DefaultReporterConfig.Verbose = true
+	// Turn on EmitSpecProgress to get spec progress (especially on interrupt)
+	ginkgoconfig.GinkgoConfig.EmitSpecProgress = true
+	// Randomize specs as well as suites
+	ginkgoconfig.GinkgoConfig.RandomizeAllSpecs = true
+
+	wait.ForeverTestTimeout = time.Second * 60
+
+	// Since we are going to be creating CRDs in the API server, we need to
+	// register these types.
+	utilruntime.Must(extapi.AddToScheme(cmpapi.GlobalScheme))
+}
+
+// RunSuite runs a Ginkgo test suite, and writes the results to the artefacts
+// directory. The artifacts directory may be overridden with `ARTIFACTS`
+// environment variable.
+func RunSuite(t *testing.T, suiteName, artifactsDir string) {
+	gomega.RegisterFailHandler(ginkgo.Fail)
+
+	if path := os.Getenv("ARTIFACTS"); len(path) > 0 {
+		artifactsDir = path
+	}
+
+	junitReporter := reporters.NewJUnitReporter(filepath.Join(
+		artifactsDir,
+		fmt.Sprintf("junit-go-%s.xml", suiteName),
+	))
+
+	ginkgo.RunSpecsWithDefaultAndCustomReporters(t, suiteName, []ginkgo.Reporter{junitReporter})
+}
+
+// RunControlPlane runs a local API server and makes it ready for running tests
+// against. Also runs the cert-manager webhook for operating over cert-manager
+// resources. Expects CRD directories to both cert-manager, as well as
+// policy-approver. This *MUST* be provided.
+// Returns a controller-runtime envtest which is ready to be run against.
+func RunControlPlane(t *testing.T, crdDirs ...string) *envtest.Environment {
+	env := &envtest.Environment{
+		AttachControlPlaneOutput: false,
+	}
+
+	t.Logf("starting API server...")
+	if _, err := env.Start(); err != nil {
+		t.Fatalf("failed to start control plane: %v", err)
+	}
+	t.Logf("running API server at %q", env.Config.Host)
+
+	// Register cleanup func to stop the api-server after the test has finished.
+	t.Cleanup(func() {
+		t.Log("stopping API server")
+		if err := env.Stop(); err != nil {
+			t.Fatalf("failed to shut down control plane: %v", err)
+		}
+	})
+
+	t.Log("writing cert-manager webhook")
+	kubeconifgPath := writeKubeconfig(t, env.Config, "cert-manager-webhook-kubeconfig.yaml")
+	t.Logf("cert-manager webhook written to %q", kubeconifgPath)
+
+	webhookOpts, stopWebhook := webhooktesting.StartWebhookServer(t, []string{"--kubeconfig=" + kubeconifgPath})
```

Comment:
Potentially.. though ideally CertificateRequests won't be changing at all post v1.3. If they do, we should be able to catch any differences by human checking since a field being added will need to add support to policy-approver anyway.